### ISSUE TITLE
Wonder the difference between morph-token and abstr-abstr(VQ) in ablation study

### ISSUE THREAD

**Issue by easonyang1996** [DCDmllm/MorphTokens#1]
abstr-abstr(VQ) also adopts a decoder-only transformers to transform post-MLLM token to complete visual token, just like the proposed morph-token. So what's the difference? Is it that the abstr-abstr(VQ) requires the pre- and post-LLM tokens to be the same?

**Comment by yangdongchao** [comment]
> abstr-abstr(VQ) also adopts a decoder-only transformers to transform post-MLLM token to complete visual token, just like the proposed morph-token. So what's the difference? Is it that the abstr-abstr(VQ) requires the pre- and post-LLM tokens to be the same?

This paper seems exists some problems. In my view, the author want to say, previous work directly use abstr (vq) as both input and output will limit the performance of generation, furthermore it also limits the performance of complex understanding task. In fact, author want to say in the formultaion of abstr-abstr(VQ), abstr (VQ) as input still includes too much information, similarly, abstr(VQ) as output still lacking enough information.  From my understanding, the output of LLM is not the token, the author just use the output embedding, and input this into a new decoder-only transformer. From formula (6), it is joint trained with LLM.  Maybe the real improvement from the joint training?
In summary, this paper exists a lot of problems. From the paper and the open-sourced code, we cannot fullly understand why the proposed method can work and how to reproduce.  If the author without any update, I do not suggest to follow. It will cost a lot of time.

**Comment by Caixy1113** [comment]
> > abstr-abstr(VQ) also adopts a decoder-only transformers to transform post-MLLM token to complete visual token, just like the proposed morph-token. So what's the difference? Is it that the abstr-abstr(VQ) requires the pre- and post-LLM tokens to be the same?
> 
> This paper seems exists some problems. In my view, the author want to say, previous work directly use abstr (vq) as both input and output will limit the performance of generation, furthermore it also limits the performance of complex understanding task. In fact, author want to say in the formultaion of abstr-abstr(VQ), abstr (VQ) as input still includes too much information, similarly, abstr(VQ) as output still lacking enough information. From my understanding, the output of LLM is not the token, the author just use the output embedding, and input this into a new decoder-only transformer. From formula (6), it is joint trained with LLM. Maybe the real improvement from the joint training? In summary, this paper exists a lot of problems. From the paper and the open-sourced code, we cannot fullly understand why the proposed method can work and how to reproduce. If the author without any update, I do not suggest to follow. It will cost a lot of time.

I believe you are right, what confuses me is how the author implemented formula 6 in the code. If they are not using the standard next token prediction loss, I wonder how the author maintains the gradients of the sampled tokens in formula 6. This should require some special techniques, which I think is the key part of the paper, but the author did not mention it at all.
<img width="611" alt="image" src="https://github.com/user-attachments/assets/e3fb8f99-dcbe-4324-aedb-fb1ecebe2c48">

**Comment by zhkzly** [comment]
I believe that in this formula, when the data is merely text, the morph token should be a placeholder such as <img>, and during prediction, a specified number of morph tokens (an improved version) should be generated to be directly used as input for the next decoder step. This would allow for training in a next-token fashion. When the input consists of both text and an image, the morph tokens generated during the process would be specific morph tokens representing the image.

![图片](https://github.com/user-attachments/assets/775d9c6c-2060-4263-bd7d-3ff67ccd3e6b)

The order of text tokens and morph tokens should be incorrect; the paper also states that text tokens should come before morph tokens. The picture may be wrong

**Comment by zhkzly** [comment]
I believe that in this formula, when the data is merely text, the morph token should be a placeholder such as<img> , and during prediction, a specified number of morph tokens (an improved version) should be generated to be directly used as input for the next decoder step. This would allow for training in a next-token fashion. When the input consists of both text and an image, the morph tokens generated during the process would be specific morph tokens representing the image.
![图片](https://github.com/user-attachments/assets/e8266063-8543-4b99-bd71-2e9f9fc2e464)
The order of text tokens and morph tokens should be incorrect; the paper also states that text tokens should come before morph tokens.

**Comment by zhkzly** [comment]
I believe that in this formula, when the data is merely text, the morph token should be a placeholder such as "<img>", and during prediction, a specified number of morph tokens (an improved version) should be generated to be directly used as input for the next decoder step. This would allow for training in a next-token fashion. When the input consists of both text and an image, the text tokens should come before the morph tokens, with the morph tokens being specific representations of the image.
![图片](https://github.com/user-attachments/assets/a4393176-691e-4ffa-96f5-21d7c5051bc2)
The order of text tokens and morph tokens should be incorrect; the paper also states that text tokens should come before morph tokens.

**Comment by zhkzly** [comment]
I think that in this formula, when the data is merely text, the morph token should be a placeholder such as "IMG", and during prediction, a specified number of morph tokens (an improved version) should be generated to be directly used as input for the next decoder step. This would allow for training in a next-token fashion. When the input consists of both text and an image, the text tokens should come before the morph tokens, with the morph tokens being specific representations of the image.
![图片](https://github.com/user-attachments/assets/4b803498-852d-4c63-936d-db58688f2bea)
The order of text tokens and morph tokens should be incorrect; the paper also states that text tokens should come before morph tokens.

**Comment by zhifeijidaxia** [comment]
I really hope authors can provide more codes to help us to repeat it，I can't understand what difference is between the morphs tokens with VQ\SD until now.
